# Cleaning the Star Wars Data

![Cleaning Star Wars](../portfolio/star_wars_cleaning3.jpg)

Hello I will be demonstrating data cleaning.


## Table of Contents
* Introduction
* Importing the Data
* Get Familiarized with the Data


## 1. Introduction

While waiting for [Star Wars: The Force Awakens](https://en.wikipedia.org/wiki/Star_Wars:_The_Force_Awakens) to come out, the team at [FiveThirtyEight](https://fivethirtyeight.com/) became interested in answering some questions about Star Wars fans. In particular, they wondered: **does the rest of America realize that “The Empire Strikes Back” is clearly the best of the bunch?**

The team needed to collect data addressing this question. To do this, they surveyed Star Wars fans using the online tool SurveyMonkey. They received 835 total responses, which you can explore for yourself from [their GitHub repository.](https://github.com/fivethirtyeight/data/tree/master/star-wars-survey)

In this project, I will provide an example of cleaning and slightly analyzing the data on the Star Wars movies.


## 2. Importing the Data

We will be using the `Pandas` library as the main tool for getting familiarized and tidying the data. Fortunately, the Pandas library also provides excellent data importing options with various essential paramters (e.g. the type of encoding). The following code will read the data from a __*"Comma-separated values"*__ (CSV) file into a pandas dataframe:

In [1]:
import pandas as pd
star_wars = pd.read_csv("star_wars.csv", encoding="ISO-8859-1")

We need to specify an `encoding` because the data set has some characters that aren't in Python's default `utf-8` encoding.


## 3. Exploring the Columns and Rows

As we now have all the data in the `star_wars` dataframe, lets begin to explore the data by first getting an overview of the column names, and also the number of columns and rows. We will obtain this information by using the `info` function in the Pandas library.

Furthermore, we will also be using the `head` and `tail` function to get some insight about the first 5 and last 5 rows in the dataframe.

In [2]:
star_wars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1187 entries, 0 to 1186
Data columns (total 38 columns):
RespondentID                                                                                                                                     1186 non-null float64
Have you seen any of the 6 films in the Star Wars franchise?                                                                                     1187 non-null object
Do you consider yourself to be a fan of the Star Wars film franchise?                                                                            837 non-null object
Which of the following Star Wars films have you seen? Please select all that apply.                                                              674 non-null object
Unnamed: 4                                                                                                                                       572 non-null object
Unnamed: 5                                                       

In [3]:
star_wars.head()

RespondentID Have you seen any of the 6 films in the Star Wars franchise?  \
0           NaN                                           Response             
1  3.292880e+09                                                Yes             
2  3.292880e+09                                                 No             
3  3.292765e+09                                                Yes             
4  3.292763e+09                                                Yes             

  Do you consider yourself to be a fan of the Star Wars film franchise?  \
0                                           Response                      
1                                                Yes                      
2                                                NaN                      
3                                                 No                      
4                                                Yes                      

  Which of the following Star Wars films have you seen? Please select all that apply.  \
0           Star Wars: Episode I  The Phantom Menace                                    
1           Star Wars: Episode I  The Phantom Menace                                    
2                                                NaN                                    
3           Star Wars: Episode I  The Phantom Menace                                    
4           Star Wars: Episode I  The Phantom Menace                                    

                                    Unnamed: 4  \
0  Star Wars: Episode II  Attack of the Clones   
1  Star Wars: Episode II  Attack of the Clones   
2                                          NaN   
3  Star Wars: Episode II  Attack of the Clones   
4  Star Wars: Episode II  Attack of the Clones   

                                    Unnamed: 5  \
0  Star Wars: Episode III  Revenge of the Sith   
1  Star Wars: Episode III  Revenge of the Sith   
2                                          NaN   
3  Star Wars: Episode III  Revenge of the Sith   
4  Star Wars: Episode III  Revenge of the Sith   

                          Unnamed: 6  \
0  Star Wars: Episode IV  A New Hope   
1  Star Wars: Episode IV  A New Hope   
2                                NaN   
3                                NaN   
4  Star Wars: Episode IV  A New Hope   

                                     Unnamed: 7  \
0  Star Wars: Episode V The Empire Strikes Back   
1  Star Wars: Episode V The Empire Strikes Back   
2                                           NaN   
3                                           NaN   
4  Star Wars: Episode V The Empire Strikes Back   

                                 Unnamed: 8  \
0  Star Wars: Episode VI Return of the Jedi   
1  Star Wars: Episode VI Return of the Jedi   
2                                       NaN   
3                                       NaN   
4  Star Wars: Episode VI Return of the Jedi   

  Please rank the Star Wars films in order of preference with 1 being your favorite film in the franchise and 6 being your least favorite film.  \
0           Star Wars: Episode I  The Phantom Menace                                                                                              
1                                                  3                                                                                              
2                                                NaN                                                                                              
3                                                  1                                                                                              
4                                                  5                                                                                              

            ...                  Unnamed: 28  \
0           ...                         Yoda   
1           ...               Very favorably   
2           ...                          NaN   
3           ...             

In [4]:
star_wars.tail()

RespondentID  \
1182  3.288389e+09   
1183  3.288379e+09   
1184  3.288375e+09   
1185  3.288373e+09   
1186  3.288373e+09   

     Have you seen any of the 6 films in the Star Wars franchise?  \
1182                                                Yes             
1183                                                Yes             
1184                                                 No             
1185                                                Yes             
1186                                                Yes             

     Do you consider yourself to be a fan of the Star Wars film franchise?  \
1182                                                Yes                      
1183                                                Yes                      
1184                                                NaN                      
1185                                                Yes                      
1186                                                 No                      

     Which of the following Star Wars films have you seen? Please select all that apply.  \
1182           Star Wars: Episode I  The Phantom Menace                                    
1183           Star Wars: Episode I  The Phantom Menace                                    
1184                                                NaN                                    
1185           Star Wars: Episode I  The Phantom Menace                                    
1186           Star Wars: Episode I  The Phantom Menace                                    

                                       Unnamed: 4  \
1182  Star Wars: Episode II  Attack of the Clones   
1183  Star Wars: Episode II  Attack of the Clones   
1184                                          NaN   
1185  Star Wars: Episode II  Attack of the Clones   
1186  Star Wars: Episode II  Attack of the Clones   

                                       Unnamed: 5  \
1182  Star Wars: Episode III  Revenge of the Sith   
1183  Star Wars: Episode III  Revenge of the Sith   
1184                                          NaN   
1185  Star Wars: Episode III  Revenge of the Sith   
1186                                          NaN   

                             Unnamed: 6  \
1182  Star Wars: Episode IV  A New Hope   
1183  Star Wars: Episode IV  A New Hope   
1184                                NaN   
1185  Star Wars: Episode IV  A New Hope   
1186                                NaN   

                                        Unnamed: 7  \
1182  Star Wars: Episode V The Empire Strikes Back   
1183  Star Wars: Episode V The Empire Strikes Back   
1184                                           NaN   
1185  Star Wars: Episode V The Empire Strikes Back   
1186  Star Wars: Episode V The Empire Strikes Back   

                                    Unnamed: 8  \
1182  Star Wars: Episode VI Return of the Jedi   
1183  Star Wars: Episode VI Return of the Jedi   
1184                                       NaN   
1185  Star Wars: Episode VI Return of the Jedi   
1186  Star Wars: Episode VI Return of the Jedi   

     Please rank the Star Wars films in order of preference with 1 being your favorite film in the franchise and 6 being your least favorite film.  \
1182                                                  5                                                                                              
1183                                                  4                                                                                              
1184                                                NaN                                                                                              
1185                                                  4                                                                                              
1186                                                  6                                                                                              

            

The `info` function in combination with `head` and `tail` are increatdibly useful as it have now provided us with the following information:
 * There are **38 columns**
 * There are **1187 rows** (including header)
 * There is only 1 `float` (ID column) and the remaining columns are the `object` type
 * There are various amount of missing values in each column
 * The naming of the columns are messy with many **Unnamed** columns and a few very long column names
 
Based on this exploartion we will need to look into the naming of the columns and also the missing values. We will start off by digging into the renaming part so it will be more intuitive to work with the dataframe.



## 4. Renaming the Columns

In order to rename the columns we will need to know specifically what each column represents. The `head` and `tail` function provided some insight to this. 

The data has several columns, including:

* **_RespondentID_** - An anonymized ID for the respondent (person taking the survey)
* **_Gender_** - The respondent's gender
* **_Age_** - The respondent's age
* **_Household Income_** - The respondent's income
* **_Education_** - The respondent's education level
* **_Location (Census Region)_** - The respondent's location
* **_Do you consider yourself to be a fan of the Star Wars film franchise?_** - Has a Yes or No response
* **_Have you seen any of the 6 films in the Star Wars franchise?_** - Has a Yes or No response

There are several other columns containing answers to questions about the Star Wars movies. For some questions, the respondent had to check one or more boxes. This type of data is difficult to represent in columnar format. As a result, many of the columns have been named **unnamed**.

The unnamed columns are basicaly an extension of the question that proceeded a sequence of unnamed columns. In order to clarify, I will illustrate the first sequence of unnamed columns: 

* **_Which of the following Star Wars films have you seen? Please select all that apply._** - Whether or not the respondent saw _Star Wars: Episode I The Phantom Menace_.
* **_Unnamed: 4_** - Whether or not the respondent saw _Star Wars: Episode II Attack of the Clones_.
* **_Unnamed: 5_** - Whether or not the respondent saw _Star Wars: Episode III Revenge of the Sith_.
* **_Unnamed: 6_** - Whether or not the respondent saw _Star Wars: Episode IV A New Hope_.
* **_Unnamed: 7_** - Whether or not the respondent saw _Star Wars: Episode V The Empire Strikes Back_.
* **_Unnamed: 8_** - Whether or not the respondent saw _Star Wars: Episode VI Return of the Jedi_.

For each of these columns, if the value in a cell is the name of the movie, that means the respondent saw the movie. If the value is `NaN`, the respondent either didn't answer or didn't see the movie. We'll assume that they didn't see the movie.

### 4.1 Renaming Unnamed 4-8 Columns
We will start off by renaming these columns to a more understandable naming. We will do this by creating a dictionary which will contain the translation of the new column names. Afterwards we will feed the `rename` funciton with the dicitonary.

In [5]:
new_names={
    'Which of the following Star Wars films have you seen? Please select all that apply.': 'seen_1',
    'Unnamed: 4': 'seen_2',
    'Unnamed: 5': 'seen_3',
    'Unnamed: 6': 'seen_4',
    'Unnamed: 7': 'seen_5',
    'Unnamed: 8': 'seen_6'}

star_wars = star_wars.rename(columns=new_names)
star_wars.iloc[0:5,0:9]

RespondentID Have you seen any of the 6 films in the Star Wars franchise?  \
0           NaN                                           Response             
1  3.292880e+09                                                Yes             
2  3.292880e+09                                                 No             
3  3.292765e+09                                                Yes             
4  3.292763e+09                                                Yes             

  Do you consider yourself to be a fan of the Star Wars film franchise?  \
0                                           Response                      
1                                                Yes                      
2                                                NaN                      
3                                                 No                      
4                                                Yes                      

                                     seen_1  \
0  Star Wars: Episode I  The Phantom Menace   
1  Star Wars: Episode I  The Phantom Menace   
2                                       NaN   
3  Star Wars: Episode I  The Phantom Menace   
4  Star Wars: Episode I  The Phantom Menace   

                                        seen_2  \
0  Star Wars: Episode II  Attack of the Clones   
1  Star Wars: Episode II  Attack of the Clones   
2                                          NaN   
3  Star Wars: Episode II  Attack of the Clones   
4  Star Wars: Episode II  Attack of the Clones   

                                        seen_3  \
0  Star Wars: Episode III  Revenge of the Sith   
1  Star Wars: Episode III  Revenge of the Sith   
2                                          NaN   
3  Star Wars: Episode III  Revenge of the Sith   
4  Star Wars: Episode III  Revenge of the Sith   

                              seen_4  \
0  Star Wars: Episode IV  A New Hope   
1  Star Wars: Episode IV  A New Hope   
2                                NaN   
3                                NaN   
4  Star Wars: Episode IV  A New Hope   

                                         seen_5  \
0  Star Wars: Episode V The Empire Strikes Back   
1  Star Wars: Episode V The Empire Strikes Back   
2                                           NaN   
3                                           NaN   
4  Star Wars: Episode V The Empire Strikes Back   

                                     seen_6  
0  Star Wars: Episode VI Return of the Jedi  
1  Star Wars: Episode VI Return of the Jedi  
2                                       NaN  
3                                       NaN  
4  Star Wars: Episode VI Return of the Jedi

### 4.2 Renaming Unnamed 10-14 and 16-28 Columns
The sequence of unnamed 4-8 have successfully been renamed to **seen_x**. We will now do this for the remaining unnamed columns. 

In [6]:
star_wars.iloc[0:5,9:15]

Please rank the Star Wars films in order of preference with 1 being your favorite film in the franchise and 6 being your least favorite film.  \
0           Star Wars: Episode I  The Phantom Menace                                                                                              
1                                                  3                                                                                              
2                                                NaN                                                                                              
3                                                  1                                                                                              
4                                                  5                                                                                              

                                   Unnamed: 10  \
0  Star Wars: Episode II  Attack of the Clones   
1                                            2   
2                                          NaN   
3                                            2   
4                                            6   

                                   Unnamed: 11  \
0  Star Wars: Episode III  Revenge of the Sith   
1                                            1   
2                                          NaN   
3                                            3   
4                                            1   

                         Unnamed: 12  \
0  Star Wars: Episode IV  A New Hope   
1                                  4   
2                                NaN   
3                                  4   
4                                  2   

                                    Unnamed: 13  \
0  Star Wars: Episode V The Empire Strikes Back   
1                                             5   
2                                           NaN   
3                                             5   
4                                             4   

                                Unnamed: 14  
0  Star Wars: Episode VI Return of the Jedi  
1                                         6  
2                                       NaN  
3                                         6  
4                                         3

In [7]:
new_names={
    'Please rank the Star Wars films in order of preference with 1 being your favorite film in the franchise and 6 being your least favorite film.': 'rank_1',
    'Unnamed: 10': 'rank_2',
    'Unnamed: 11': 'rank_3',
    'Unnamed: 12': 'rank_4',
    'Unnamed: 13': 'rank_5',
    'Unnamed: 14': 'rank_6'}

star_wars = star_wars.rename(columns=new_names)
star_wars.iloc[0:5,9:15]

rank_1  \
0  Star Wars: Episode I  The Phantom Menace   
1                                         3   
2                                       NaN   
3                                         1   
4                                         5   

                                        rank_2  \
0  Star Wars: Episode II  Attack of the Clones   
1                                            2   
2                                          NaN   
3                                            2   
4                                            6   

                                        rank_3  \
0  Star Wars: Episode III  Revenge of the Sith   
1                                            1   
2                                          NaN   
3                                            3   
4                                            1   

                              rank_4  \
0  Star Wars: Episode IV  A New Hope   
1                                  4   
2                                NaN   
3                                  4   
4                                  2   

                                         rank_5  \
0  Star Wars: Episode V The Empire Strikes Back   
1                                             5   
2                                           NaN   
3                                             5   
4                                             4   

                                     rank_6  
0  Star Wars: Episode VI Return of the Jedi  
1                                         6  
2                                       NaN  
3                                         6  
4                                         3

In [8]:
star_wars.iloc[0:5,15:29]

Please state whether you view the following characters favorably, unfavorably, or are unfamiliar with him/her.  \
0                                           Han Solo                                                               
1                                     Very favorably                                                               
2                                                NaN                                                               
3                                 Somewhat favorably                                                               
4                                     Very favorably                                                               

          Unnamed: 16           Unnamed: 17         Unnamed: 18  \
0      Luke Skywalker  Princess Leia Organa    Anakin Skywalker   
1      Very favorably        Very favorably      Very favorably   
2                 NaN                   NaN                 NaN   
3  Somewhat favorably    Somewhat favorably  Somewhat favorably   
4      Very favorably        Very favorably      Very favorably   

          Unnamed: 19         Unnamed: 20       Unnamed: 21  \
0      Obi Wan Kenobi   Emperor Palpatine       Darth Vader   
1      Very favorably      Very favorably    Very favorably   
2                 NaN                 NaN               NaN   
3  Somewhat favorably    Unfamiliar (N/A)  Unfamiliar (N/A)   
4      Very favorably  Somewhat favorably    Very favorably   

          Unnamed: 22           Unnamed: 23       Unnamed: 24  \
0    Lando Calrissian             Boba Fett             C-3P0   
1    Unfamiliar (N/A)      Unfamiliar (N/A)    Very favorably   
2                 NaN                   NaN               NaN   
3    Unfamiliar (N/A)      Unfamiliar (N/A)  Unfamiliar (N/A)   
4  Somewhat favorably  Somewhat unfavorably    Very favorably   

        Unnamed: 25       Unnamed: 26       Unnamed: 27       Unnamed: 28  
0             R2 D2     Jar Jar Binks     Padme Amidala              Yoda  
1    Very favorably    Very favorably    Very favorably    Very favorably  
2               NaN               NaN               NaN               NaN  
3  Unfamiliar (N/A)  Unfamiliar (N/A)  Unfamiliar (N/A)  Unfamiliar (N/A)  
4    Very favorably    Very favorably    Very favorably    Very favorably

In [9]:
new_names={
    'Please state whether you view the following characters favorably, unfavorably, or are unfamiliar with him/her.': 'Han Solo',
    'Unnamed: 16': 'Luke Skywalker',
    'Unnamed: 17': 'Princess Leia Organa',
    'Unnamed: 18': 'Anakin Skywalker',
    'Unnamed: 19': 'Obi Wan Kenobi',
    'Unnamed: 20': 'Emperor Palpatine',
    'Unnamed: 21': 'Darth Vader',
    'Unnamed: 22': 'Lando Calrissian',
    'Unnamed: 23': 'Boba Fett',
    'Unnamed: 24': 'C-3P0',
    'Unnamed: 25': 'R2 D2',
    'Unnamed: 26': 'Jar Jar Binks',
    'Unnamed: 27': 'Padme Amidala',
    'Unnamed: 28': 'Yoda'}

star_wars = star_wars.rename(columns=new_names)
star_wars.iloc[0:5,15:29]

Han Solo      Luke Skywalker  Princess Leia Organa  \
0            Han Solo      Luke Skywalker  Princess Leia Organa   
1      Very favorably      Very favorably        Very favorably   
2                 NaN                 NaN                   NaN   
3  Somewhat favorably  Somewhat favorably    Somewhat favorably   
4      Very favorably      Very favorably        Very favorably   

     Anakin Skywalker      Obi Wan Kenobi   Emperor Palpatine  \
0    Anakin Skywalker      Obi Wan Kenobi   Emperor Palpatine   
1      Very favorably      Very favorably      Very favorably   
2                 NaN                 NaN                 NaN   
3  Somewhat favorably  Somewhat favorably    Unfamiliar (N/A)   
4      Very favorably      Very favorably  Somewhat favorably   

        Darth Vader    Lando Calrissian             Boba Fett  \
0       Darth Vader    Lando Calrissian             Boba Fett   
1    Very favorably    Unfamiliar (N/A)      Unfamiliar (N/A)   
2               NaN                 NaN                   NaN   
3  Unfamiliar (N/A)    Unfamiliar (N/A)      Unfamiliar (N/A)   
4    Very favorably  Somewhat favorably  Somewhat unfavorably   

              C-3P0             R2 D2     Jar Jar Binks     Padme Amidala  \
0             C-3P0             R2 D2     Jar Jar Binks     Padme Amidala   
1    Very favorably    Very favorably    Very favorably    Very favorably   
2               NaN               NaN               NaN               NaN   
3  Unfamiliar (N/A)  Unfamiliar (N/A)  Unfamiliar (N/A)  Unfamiliar (N/A)   
4    Very favorably    Very favorably    Very favorably    Very favorably   

               Yoda  
0              Yoda  
1    Very favorably  
2               NaN  
3  Unfamiliar (N/A)  
4    Very favorably

### 4.3 Inspecting the Renamed Columns
Now that we have renamed all the unnamed columns, lets get an overview of all the names in a list format.

In [10]:
star_wars.columns

Index(['RespondentID',
       'Have you seen any of the 6 films in the Star Wars franchise?',
       'Do you consider yourself to be a fan of the Star Wars film franchise?',
       'seen_1', 'seen_2', 'seen_3', 'seen_4', 'seen_5', 'seen_6', 'rank_1',
       'rank_2', 'rank_3', 'rank_4', 'rank_5', 'rank_6', 'Han Solo',
       'Luke Skywalker', 'Princess Leia Organa', 'Anakin Skywalker',
       'Obi Wan Kenobi', 'Emperor Palpatine', 'Darth Vader',
       'Lando Calrissian', 'Boba Fett', 'C-3P0', 'R2 D2', 'Jar Jar Binks',
       'Padme Amidala', 'Yoda', 'Which character shot first?',
       'Are you familiar with the Expanded Universe?',
       'Do you consider yourself to be a fan of the Expanded Universe???',
       'Do you consider yourself to be a fan of the Star Trek franchise?',
       'Gender', 'Age', 'Household Income', 'Education',
       'Location (Census Region)'],
      dtype='object')

## 5. Removing Missing Data

The columns have now been renamed to a naming which is more intuitive to understand. Now we will explore the amount of missing data. We can do this by using the `isnull` in combination with the `sum` functions.

In [11]:
star_wars.isnull().sum()

RespondentID                                                               1
Have you seen any of the 6 films in the Star Wars franchise?               0
Do you consider yourself to be a fan of the Star Wars film franchise?    350
seen_1                                                                   513
seen_2                                                                   615
seen_3                                                                   636
seen_4                                                                   579
seen_5                                                                   428
seen_6                                                                   448
rank_1                                                                   351
rank_2                                                                   350
rank_3                                                                   351
rank_4                                                                   350

From this initial inspection, it could seem that there is an unreasonably high amount of missing data. However, we will soon see that this may not be the case for most columns in the dataframe. We will get into this in the next section.

Before we move on, we will need to remove the row with a missing value in the respondentID column.

In [12]:
star_wars = star_wars[pd.notnull(star_wars['RespondentID'])]
star_wars['RespondentID'].isnull().sum()

0

## 6. Mapping Values

In the previous section we observed a high number of missing values across all columns. But when we begin to investigate this pattern we will see a quite another explanation for the missing values.

In [13]:
star_wars.iloc[0:5,0:9]

RespondentID Have you seen any of the 6 films in the Star Wars franchise?  \
1  3.292880e+09                                                Yes             
2  3.292880e+09                                                 No             
3  3.292765e+09                                                Yes             
4  3.292763e+09                                                Yes             
5  3.292731e+09                                                Yes             

  Do you consider yourself to be a fan of the Star Wars film franchise?  \
1                                                Yes                      
2                                                NaN                      
3                                                 No                      
4                                                Yes                      
5                                                Yes                      

                                     seen_1  \
1  Star Wars: Episode I  The Phantom Menace   
2                                       NaN   
3  Star Wars: Episode I  The Phantom Menace   
4  Star Wars: Episode I  The Phantom Menace   
5  Star Wars: Episode I  The Phantom Menace   

                                        seen_2  \
1  Star Wars: Episode II  Attack of the Clones   
2                                          NaN   
3  Star Wars: Episode II  Attack of the Clones   
4  Star Wars: Episode II  Attack of the Clones   
5  Star Wars: Episode II  Attack of the Clones   

                                        seen_3  \
1  Star Wars: Episode III  Revenge of the Sith   
2                                          NaN   
3  Star Wars: Episode III  Revenge of the Sith   
4  Star Wars: Episode III  Revenge of the Sith   
5  Star Wars: Episode III  Revenge of the Sith   

                              seen_4  \
1  Star Wars: Episode IV  A New Hope   
2                                NaN   
3                                NaN   
4  Star Wars: Episode IV  A New Hope   
5  Star Wars: Episode IV  A New Hope   

                                         seen_5  \
1  Star Wars: Episode V The Empire Strikes Back   
2                                           NaN   
3                                           NaN   
4  Star Wars: Episode V The Empire Strikes Back   
5  Star Wars: Episode V The Empire Strikes Back   

                                     seen_6  
1  Star Wars: Episode VI Return of the Jedi  
2                                       NaN  
3                                       NaN  
4  Star Wars: Episode VI Return of the Jedi  
5  Star Wars: Episode VI Return of the Jedi

### 6.1 Mapping Yes/No Columns

Take a look at the next two columns after RespondentID, which are:

* ___Have you seen any of the 6 films in the Star Wars franchise?___
* ___Do you consider yourself to be a fan of the Star Wars film franchise?___

Both represent _Yes/No_ questions. They can also be _NaN_ where a respondent chooses not to answer a question. We can use the *pandas.Series.value\_counts()* method on a series to see all of the unique values in a column, along with the total number of times each value appears.

In [14]:
star_wars['Have you seen any of the 6 films in the Star Wars franchise?'].value_counts()

Yes    936
No     250
Name: Have you seen any of the 6 films in the Star Wars franchise?, dtype: int64

In [15]:
star_wars['Do you consider yourself to be a fan of the Star Wars film franchise?'].value_counts()

Yes    552
No     284
Name: Do you consider yourself to be a fan of the Star Wars film franchise?, dtype: int64

Both columns are currently _string types_, because the main values they contain are _Yes_ and _No_. We can make the data a bit easier to analyze down the road by converting each column to a _Boolean_ having only the values True, False, and NaN. Booleans are easier to work with because we can select the rows that are True or False without having to do a string comparison.

We can use the _pandas.Series.map()_ method on series objects to perform the conversion.

As we can see in the first 5 rows of the dataframe, missing values may indicate the answer "no" whereas the name of the star wars movie represents the answer "yes". For the purpose of this project we will assume the missing values in these columns indeed represents the answer "no". Thus, we will  

In [16]:
yes_no = {
    "Yes": True,
    "No": False
}
star_wars['Have you seen any of the 6 films in the Star Wars franchise?'] = star_wars['Have you seen any of the 6 films in the Star Wars franchise?'].map(yes_no)
star_wars['Have you seen any of the 6 films in the Star Wars franchise?'].value_counts()

True     936
False    250
Name: Have you seen any of the 6 films in the Star Wars franchise?, dtype: int64

In [17]:
star_wars['Do you consider yourself to be a fan of the Star Wars film franchise?'] = star_wars['Do you consider yourself to be a fan of the Star Wars film franchise?'].map(yes_no)
star_wars['Do you consider yourself to be a fan of the Star Wars film franchise?'].value_counts()

True     552
False    284
Name: Do you consider yourself to be a fan of the Star Wars film franchise?, dtype: int64

We can see the value_counts demonstrated that the 2 columns have indeed been converted to boolean. I will continue to do the same for the 3 remaning yes/nocolumns in the dataframe.

In [18]:
star_wars['Are you familiar with the Expanded Universe?'] = star_wars['Are you familiar with the Expanded Universe?'].map(yes_no)
star_wars['Do you consider yourself to be a fan of the Expanded Universe???'] = star_wars['Do you consider yourself to be a fan of the Expanded Universe???'].map(yes_no)
star_wars['Do you consider yourself to be a fan of the Star Trek franchise?'] = star_wars['Do you consider yourself to be a fan of the Star Trek franchise?'].map(yes_no)

### 6.2 Mapping Checkbox Columns

The next six columns represent a single checkbox question. The respondent checked off a series of boxes in response to the question, ***Which of the following Star Wars films have you seen? Please select all that apply***.

For each of these columns, if the value in a cell is the name of the movie, that means the respondent saw the movie. If the value is ***NaN***, the respondent either didn't answer or didn't see the movie. We'll assume that they didn't see the movie.

We'll need to convert each of these columns to a ***Boolean***. We can convert the values the same way we did earlier, except that we'll need to include the movie title and NaN in the mapping dictionary.

In [19]:
import numpy as np
true_false = {
    np.NaN: False,
    "Star Wars: Episode I  The Phantom Menace": True,
    "Star Wars: Episode II  Attack of the Clones": True,
    "Star Wars: Episode III  Revenge of the Sith": True,
    "Star Wars: Episode IV  A New Hope": True,
    "Star Wars: Episode V The Empire Strikes Back": True,
    "Star Wars: Episode VI Return of the Jedi": True
}

for i in range(3,9):
    star_wars.iloc[:,i] = star_wars.iloc[:,i].map(true_false)
    
star_wars.iloc[0:5,0:9]

RespondentID  Have you seen any of the 6 films in the Star Wars franchise?  \
1  3.292880e+09                                               True              
2  3.292880e+09                                              False              
3  3.292765e+09                                               True              
4  3.292763e+09                                               True              
5  3.292731e+09                                               True              

  Do you consider yourself to be a fan of the Star Wars film franchise?  \
1                                               True                      
2                                                NaN                      
3                                              False                      
4                                               True                      
5                                               True                      

   seen_1  seen_2  seen_3  seen_4  seen_5  seen_6  
1    True    True    True    True    True    True  
2   False   False   False   False   False   False  
3    True    True    True   False   False   False  
4    True    True    True    True    True    True  
5    True    True    True    True    True    True

All the "seen" columns have successfully been mapped to boolean type.